pubnub/models/subscription.py

```diff
@@ -0,0 +1,296 @@
+                    self.channel_groups[channel_group] = [subscription]
+                    channel_list.append(channel_group)
+                else:
+                    self.channel_groups[channel_group].append(subscription)
+
+        tt = self.pubnub._subscription_manager._timetoken
+        if channel_list:
+            subscribe_operation = SubscribeOperation(
+                channels=self.get_subscribed_channels(),
+                channel_groups=self.get_subscribed_channel_groups(),
+                timetoken=tt,
+                presence_enabled=False
+            )
+            self.pubnub._subscription_manager.adapt_subscribe_builder(subscribe_operation)
+
+        return channel_list
+
+    def remove(self, subscription: Subscription) -> list:
+        channel_list = []
+        group_list = []
+
+        if subscription.subscription_type == PNSubscriptionType.CHANNEL:
+            for channel in subscription.subscription_names:
+                if channel in self.channels:
+                    self.channels[channel].remove(subscription)
+                    if self.channels[channel] == []:
+                        del self.channels[channel]
+                        channel_list.append(channel)
+        else:
+            for channel_group in subscription.subscription_names:
+                if channel_group in self.channel_groups:
+                    self.channel_groups[channel_group].remove(subscription)
+                if not self.channel_groups[channel_group]:
+                    del self.channel_groups[channel_group]
+                    group_list.append(channel_group)
+
+        if channel_list:
+            unsubscribe_operation = UnsubscribeOperation(
+                channels=channel_list,
+                channel_groups=group_list
+            )
+            self.pubnub._subscription_manager.adapt_unsubscribe_builder(unsubscribe_operation)
+        return channel_list
+
+    def get_subscribed_channels(self):
+        return list(self.channels.keys())
+
+    def get_subscribed_channel_groups(self):
+        return list(self.channel_groups.keys())
+
+    def get_subscriptions_for(self, subscription_type: PNSubscriptionType, name: str):
+        if subscription_type == PNSubscriptionType.CHANNEL:
+            return [channel for channel in self.get_subscribed_channels() if channel == name]
+        else:
+            return [group for group in self.get_subscribed_channel_groups() if group == name]
+
+    def get_all_listeners(self):
+        listeners = []
+
+        for channel in self.channels:
+            listeners += self.channels[channel]
+        for channel_group in self.channel_groups:
+            listeners += self.channel_groups[channel_group]
+        if self.global_listeners:
+            listeners += self.global_listeners
+        return set(listeners)
+
+    def add_listener(self, listener):
+        assert isinstance(listener, SubscribeCallback)
```

Comment:
I don't fully follow how listeners work here yet, but shouldn't it only accept something like EventListener that doesn't have the status method? otherwise the user might be confused why they're not getting anything in status, they shouldn't be able to define it at all

Reply:
It's for backwards compatibility. All listeners should stem from `SubscribeCallback` metaclass. Listerers add functionality like `wait_for_message_on(channel)` but in general still use the interface described in `SubscribeCallback`